docs/using-the-compiler.rst

```diff
@@ -0,0 +1,249 @@
+there is a standardized way for describing the relations between those files.
+Furthermore, the compiler can generate a json file while compiling that includes
+the (hash of the) source, natspec comments and other metadata whose hash is included in the
+actual bytecode. Specifically, the creation data for a contract has to begin with
+`push32 <metadata hash> pop`.
+
+The metadata standard is versioned. Future versions are only required to provide the "version" field,
+the "language" field and the two keys inside the "compiler" field.
+The field compiler.keccak should be the keccak hash of a binary of the compiler with the given version.
+
+The example below is presented in a human-readable way. Properly formatted metadata
+should use quotes correctly, reduce whitespace to a minimum and sort the keys of all objects
+to arrive at a unique formatting.
+
+Comments are of course not permitted and used here only for explanatory purposes.
+
+Input Description
+-----------------
+
+QUESTION: How to specific file-reading callback? - probably not as part of json input
+
+The input description is language-specific and could change with each compiler version, but it
+should be backwards compatible if possible.
+
+    {
+      sources:
+      {
+        // the keys here are the "global" names of the source files, imports can use other files via remappings (see below)
+        "abc": "contract b{}", // specify source directly
+        // (axic) I think 'keccak' on its on is not enough. I would go perhaps with swarm: "0x12.." and ipfs: "Qma..." for simplicity
+        // (chriseth) Where the content is stored is a second component, but yes, we could give an indication there.
+        "def": {keccak: "0x123..."}, // source has to be retrieved by its hash
+        "ghi": {file: "/tmp/path/to/file.sol"}, // file on filesystem
+        // (axic) I'm inclined to think the source _must_ be provided in the JSON,
+
+        "dir/file.sol": "contract a {}"
+      },
+      settings:
+      {
+        remappings: [":g/dir"], // just as it used to be
+        // (axic) what is remapping doing exactly?
+        optimizer: {enabled: true, runs: 500},
+        // if given, only compiles this contract, can also be an array. If only a contract name is given, tries to find it if unique.
+        compilationTarget: "myFile.sol:MyContract",
+        // addresses of the libraries. If not all libraries are given here, it can result in unlinked objects whose output data is different
+        libraries: {
+          "def:MyLib": "0x123123..."
+        },
+        // The following can be used to restrict the fields the compiler will output.
+        // (axic)
+        outputSelection: [
+            "abi", "evm.assembly", "evm.bytecode", ..., "why3", "ewasm.wasm"
+        ]
+        outputSelection: {
+        abi,asm,ast,bin,bin-runtime,clone-bin,devdoc,interface,opcodes,srcmap,srcmap-runtime,userdoc
```

Comment:
Output types:
- `abi`
- `ast` (as a JSON object)
- `evm.assembly`
- ...

To be defined soon.